ath79: add support for TP-Link CPE710-v1

apcameron opened the pull request:
[oem.log](https://github.com/openwrt/openwrt/files/6312987/oem.log)
[openwrt.log](https://github.com/openwrt/openwrt/files/6312990/openwrt.log)

TP-Link CPE710-v1 is an outdoor wireless CPE for 5 GHz with
one Ethernet port based on the AP152 reference board

Specifications:
- SoC: QCA9563-AL3A MIPS 74kc @ 775MHz, AHB @ 258MHz
- RAM: 128MiB DDR2 @ 650MHz
- Flash: 16MiB SPI NOR Based on the GD25Q128
- Wi-Fi 5Ghz: ath10k chip (802.11ac for up to 867Mbps on 5GHz wireless data rate) Based on the QCA9896
- Ethernet: one 1GbE port
- 23dBi high-gain directional 2×2 MIMO antenna and a dedicated metal reflector
- Power, LAN, WLAN5G Blue LEDs
- 3x Blue LEDs

Flashing instructions:
Flash factory image through stock firmware WEB UI or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 30-40 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

apcameron committed:
ath79: add support for TP-Link CPE710-v1

TP-Link CPE710-v1 is an outdoor wireless CPE for 5 GHz with
one Ethernet port based on the AP152 reference board

Specifications:
- SoC: QCA9563-AL3A MIPS 74kc @ 775MHz, AHB @ 258MHz
- RAM: 128MiB DDR2 @ 650MHz
- Flash: 16MiB SPI NOR Based on the GD25Q128
- Wi-Fi 5Ghz: ath10k chip (802.11ac for up to 867Mbps on 5GHz wireless data rate) Based on the QCA9896
- Ethernet: one 1GbE port
- 23dBi high-gain directional 2×2 MIMO antenna and a dedicated metal reflector
- Power, LAN, WLAN5G Blue LEDs
- 3x Blue LEDs

Flashing instructions:
Flash factory image through stock firmware WEB UI or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 30-40 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron <apcameron@softhome.net>

s-2 commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -362,6 +362,16 @@ define Device/tplink_cpe610-v2
 endef
 TARGET_DEVICES += tplink_cpe610-v2
 
+define Device/tplink_cpe710-v1
+  $(Device/tplink-safeloader-uimage)
+  SOC := qca9563
+  IMAGE_SIZE := 15680k
+  DEVICE_MODEL := CPE710
+  DEVICE_VARIANT := v1
+  TPLINK_BOARD_ID := CPE710V1
```
Would we need `DEVICE_PACKAGES` for ath10k here?

apcameron replied:
It works without it but I have added it back to be consistent with the other devices.

s-2 commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,145 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+        model = "TP-Link CPE710 v1";
+        compatible = "tplink,cpe710-v1", "qca,qca9563";
+
+
+	aliases {
+		label-mac-device = &eth0;
+                led-boot = &led_lan;
+                led-failsafe = &led_lan;
+                led-upgrade = &led_lan;
+
+	};
+
+        leds {
+                compatible = "gpio-leds";
+
+                led_lan: lan {
+                        label = "blue:lan";
+                        gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+                };
+
+                wlan5g {
+                        label = "blue:wlan5g";
+                        gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+                        linux,default-trigger = "phy0tpt";
+                };
+        };
```
Tabs vs. spaces seem to be mixed here.

apcameron replied:
I have removed the spaces.

s-2 commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,145 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
```
will higher frequencies work as well, e.g. `50000000`?

apcameron replied:
I have kept it as 25000000 to be consistent with the other TP-LINK CPE Devices as well the the other devices based on the AP152 board

mpratt14 replied:
actually, max frequency for the flash is not something "consistent" based on the board model, but rather flash model. The other CPE devices have 8 MB chips. Almost all 16 MB chips are capable of at least 40 MHz

(btw it's nice to have flash chip model in commit message, you can find it in kernel log if you don't want to open it up)
(also there is public images of the inside via the FCC ID)

apcameron replied:
The TP-LINK EAP devices with a 16MB Flash also based on the AP152 Reference Board all have 25000000 as well.

mpratt14 replied:
those EAP models would also benefit from max frequency being bumped up to 40 MHz

apcameron replied:
What is the exact value I should try and is there any risk of it shortening the life of the flash?

mpratt14 replied:
there is no risk to the life of the flash here, it either reads or it doesn't

replace the "25" with "40" for 40 MHz

mpratt14 commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,142 @@
+
+	qca956x-serdes-fixup;
+
+	gmac-config {
+		device = <&gmac>;
+	};
```
gmac-config node can be removed, since you are not setting any register with it

apcameron replied:
Unfortunately removing it breaks the ethernet connectivity and I had to put it back.

mpratt14 replied:
thats interesting....i never experienced anything like that

maybe this node should be in DTSI?

mpratt14 commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,142 @@
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
```
label property should not have a space, but I would just delete the line

according to docs, if there is no label property, it takes the node name as the label (in this case, "reset")
the only reason the label property is needed is so Openwrt scripts can systematically control the LEDs

jwmullally replied:
Some context on what all the other DTS files are doing here, `label = "reset"` seems to be the current winner :)

```bash
$ grep -rh -B1 KEY_RESTART target/linux/ath79/dts | grep label | sort | uniq -c
    101             label = "reset";
      7             label = "Reset";
     50             label = "Reset button";
```

apcameron replied:
@jwmullally The CPE and WBS Devices that have it specified all use "Reset button"

adschm replied:
> according to docs, if there is no label property, it takes the node name as the label (in this case, "reset")
the only reason the label property is needed is so Openwrt scripts can systematically control the LEDs

@mpratt14 I thought this was only the case for LEDs? Can you point me to the docs in question?

mpratt14 commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,142 @@
+		led-boot = &led_lan;
+		led-failsafe = &led_lan;
+		led-upgrade = &led_lan;
+		};
```
closing bracket of aliases has an extra tab or something

apcameron commented:
@adschm  Any comments on my PR? I think this is the first TP Link CPE device based on this SOC and AC chipset.

apcameron commented:
@adschm This PR can probably be easily backported to the 21.02 release

adschm commented:
@apcameron I will certainly look at this, but I am extraordinarily short on time at the moment.

Backporting might indeed be an option, but that's something to discuss later ...

mpratt14 commented:
It would be nice to have flash model and ath10k wifi chip model in the commit message

also, what is the FCC ID? maybe the FCC pictures would help with that...

adschm commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,141 @@
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
```
Missing label.

apcameron replied:
@adschm I have added the reset label back as requested. I did have it but @mpratt14 asked me to remove it.

mpratt14 replied:
@adschm maybe this is something you would rather handle as a treewide commit to edit DTS files, like many other you have done...

according to Linux docs, the label property is deprecated and when it is missing, the node name is taken as the label

we still need labels for LEDs...unless we come up with a common node name scheme instead
but I know for sure with gpio-keys it makes no difference

adschm replied:
I'd be open to dropping label on keys, since it's much more redundant here and in contrast to LEDs, it would be actually possible to use the node name. But we first should make sure that this actually applies to keys and applies for kernel 5.4.

mpratt14 replied:
@adschm I looked deeper into the label thing and part of what I said was wrong...

for gpio-leds:
label is considered deprecated, but we currently need it
see:
https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/common.yaml

for gpio-keys:
label is not deprecated, but is used only as a "description" and is optional, and we do not need it
when label is missing, the string "gpio_keys" is the default value
however, to use the node name as the label would be a very easy patch
(converting 1 line to an if / else statement)
see lines 484 and 731 or search "button->desc":
https://elixir.bootlin.com/linux/v5.4.119/source/drivers/input/keyboard/gpio_keys.c

adschm replied:
@mpratt14 I'd support such a change, but only if it's in kernel (which we could backport then).
If you intend to submit, please put me in Cc. I'd personally frame this as achieving consistency with the behavior for LEDs, but I have not much experience myself in what the kernel people would like to hear.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -22,6 +22,9 @@ tplink,cpe610-v1|\
 tplink,cpe610-v2)
 	ucidef_set_led_netdev "lan" "LAN" "green:lan" "eth0"
 	;;
+tplink,cpe710-v1)
```
Please move according to alphabetic sorting.

adschm replied:
Update: You can merge this with the comfast,cf-wr752ac-v1|\ block

adschm commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -238,6 +238,12 @@ case "$FIRMWARE" in
 		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
 			/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
 		;;
+	tplink,cpe710-v1)
```
Sorting ...

apcameron commented:
@svanheule  The CPE710 also uses the reference board used by some of the EAP 2x5 devices. Thanks to your efforts it was easier to port this device.

svanheule commented:
Regarding the flash speed, I also looked at this when porting the other AP152-based boards:
https://lists.infradead.org/pipermail/openwrt-devel/2020-July/030236.html
https://lists.infradead.org/pipermail/openwrt-devel/2020-July/030243.html

mpratt14 commented:
IMO, 40 MHz is safe for all SOCs and all flash that is 16 MB or more

I would not recommend setting the frequency to *exactly* what the datasheet says is the maximum for normal reads, usually that value is more than 40 MHz

jwmullally commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -539,6 +539,42 @@ static struct device_info boards[] = {
+			{"log", 0xfe0000, 0x10000},
+			{"radio", 0xff0000, 0x10000},
+			{NULL, 0, 0}
+		},
```
`serial-number` is missing from partitions

```bash
grep -a "partition .*$" CPE710\(UN\)_1.0-up-ver2-2-0-P1\[20200304-rel70345\].bin
partition fs-uboot base 0x00000 size 0x50000
partition partition-table base 0x50000 size 0x02000
partition default-mac base 0x60000 size 0x00020
partition serial-number base 0x60100 size 0x00020
partition product-info base 0x61100 size 0x00100
partition device-info base 0x61400 size 0x00400
partition signature base 0x62000 size 0x00400
partition device-id base 0x63000 size 0x00100
partition os-image base 0x70000 size 0x170000
partition soft-version base 0x1e0000 size 0x00100
partition support-list base 0x1e1000 size 0x01000
partition file-system base 0x1f0000 size 0xdd0000
partition user-config base 0xfc0000 size 0x10000
partition default-config base 0xfd0000 size 0x10000
partition log base 0xfe0000 size 0x10000
partition radio base 0xff0000 size 0x10000
```

apcameron replied:
@jwmullally is you have a look at tools/firmware-utils/src/tplink-safeloader.c
You will find that it is omitted on all the CPE and WBS Devices.
The standard has been to Not include it in the definitions.

jwmullally replied:
Without `serial-number` defined, it looks like that data is being erased. /dev/mtd3 offset 0x100 shows as blank. But I can't confirm there was data there in the first place as I've now flashed my 2 units :laughing: 
```
root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00040000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "partition-table"
mtd3: 00010000 00010000 "info"
mtd4: 00f50000 00010000 "firmware"
mtd5: 00220000 00010000 "kernel"
mtd6: 00d30000 00010000 "rootfs"
mtd7: 00990000 00010000 "rootfs_data"
mtd8: 00030000 00010000 "config"
mtd9: 00010000 00010000 "art"
root@OpenWrt:~# hexdump -C /dev/mtd3
00000000  00 00 00 06 00 00 00 00  00 5f 67 03 3f 5c ff ff  |........._g.?\..|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00001100  00 00 00 40 00 00 00 00  43 50 45 37 31 30 28 54  |...@....CPE710(T|
00001110  50 2d 4c 49 4e 4b 7c 55  4e 7c 41 43 38 36 36 2d  |P-LINK|UN|AC866-|
00001120  35 7c 30 30 30 30 30 30  30 30 29 3a 31 2e 30 00  |5|00000000):1.0.|
00001130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00001140  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff ff  |................|
00001150  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
```

I suspect the TP-Link OEM flash process clears any undefined or misdefined areas ([one example](https://github.com/openwrt/openwrt/pull/2991#discussion_r444491645)), but I don't know for sure.

apcameron replied:
Here is what mine looked like before flashing it for the first time
<pre>*
00011100  00 00 00 28 00 00 00 00  43 50 45 37 31 30 28 54  |...(....CPE710(T|
00011110  50 2d 4c 49 4e 4b 7c 55  53 7c 41 43 38 36 36 2d  |P-LINK|US|AC866-|
00011120  35 7c 35 35 35 33 30 30  30 30 29 3a 31 2e 30 0a  |5|55530000):1.0.|
00011130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00011140  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff ff  |................|
00011150  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
</pre>
And Here is the Hardware ID
<pre>00011400  00 00 03 f8 00 00 00 00  68 77 5f 69 64 3a 39 33  |........hw_id:93|
00011410  42 31 42 36 34 46 46 30  36 31 35 38 36 36 30 34  |B1B64FF061586604|
00011420  35 46 42 44 43 39 45 42  38 34 42 32 39 36 0a 6f  |5FBDC9EB84B296.o|
00011430  65 6d 5f 69 64 3a 39 32  33 39 31 43 33 42 30 46  |em_id:92391C3B0F|
00011440  32 33 37 43 46 37 45 42  35 37 30 37 46 32 35 35  |237CF7EB5707F255|
00011450  30 33 46 32 41 44 0a 70  72 6f 64 75 63 74 5f 6e  |03F2AD.product_n|
00011460  61 6d 65 3a 63 70 65 37  31 30 0a 70 72 6f 64 75  |ame:cpe710.produ|
00011470  63 74 5f 76 65 72 3a 31  2e 30 0a 00 00 00 00 00  |ct_ver:1.0......|
00011480  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
</pre>

jwmullally replied:
Is there anything at absolute address 0x60100? (which looks like relative addr 0x10100 in your pasted info)?

jwmullally replied:
I just tested this by writing data to 0x60100 and flashing factory.bin with and without this line. Without it, this area is cleared during flashing.
```
{"serial-number", 0x60100, 0x00020},
```

NeoRaider replied:
I don't know about newer revisions, but the CPE210/510 1.0 that I added to OpenWrt never had a serial-number partition. The partition layout for new devices should based on TP-Link's layout for the specific device, and I agree that the serial-number partition must be preserved on CPE710.

apcameron replied:
@NeoRaider @adschm @mpratt14 @jwmullally @robimarko  I believe that all the Newer TP Link CPE/WBS devices have the Serial Number Defined in the vendor firmware but do not define it in the OpenWrt Software.
See Vendor Layout for the TP LINK CPE210 V2 below and was initially added by @robimarko and others
<pre>(CPE210(TP-LINK|UN|N300-2|00000000):2.0
fwup-ptn partition-table base 0x00800 size 0x00800      
fwup-ptn os-image base 0x01000 size 0x100001    
fwup-ptn soft-version base 0x101001 size 0x00015        
fwup-ptn support-list base 0x101016 size 0x000cd        
fwup-ptn file-system base 0x1010e3 size 0x49a001        
partition fs-uboot base 0x00000 size 0x20000
partition partition-table base 0x20000 size 0x02000
partition default-mac base 0x30000 size 0x00020
partition serial-number base 0x30100 size 0x00020
partition product-info base 0x31100 size 0x00100
partition device-info base 0x31400 size 0x00400
partition signature base 0x32000 size 0x00400
partition device-id base 0x33000 size 0x00100
partition os-image base 0x40000 size 0x170000
partition soft-version base 0x1b0000 size 0x00100
partition support-list base 0x1b1000 size 0x01000
partition file-system base 0x1c0000 size 0x600000
partition user-config base 0x7c0000 size 0x10000
partition default-config base 0x7d0000 size 0x10000
partition log base 0x7e0000 size 0x10000
partition radio base 0x7f0000 size 0x10000
</pre>
If you feel it should be defined on the CPE710 then I believe that someone needs to open a PR to address this on ALL the TP-LINK CPE/WBS Devices that have it defined so that they are consistent and then I will update this PR to match as well.

jwmullally replied:
The easiest way to ensure correct coverage there is to:

* Download all known CPE firmwares (while being friendly to the HTTP servers)
* Extract the partitions list
* Get all make/models with `serial-number` defined

(Similar to https://github.com/jwmullally/openwrt_stuff/blob/master/tl-wpa8630-v2/report/report.tsv but with different columns).

That should produce a clear list of devices to update.

(You could also check for all partitions, but it seems like this is the only one missing).

If i have time over the next week I can produce that report and/or PR.

apcameron replied:
@jwmullally I have added serial-number for the TP-LINK CPE710. Please test again with your device.

jwmullally replied:
Ok, looks like that fixed the serial issue. Tested OEM -> factory.bin and sysupgrades.

jwmullally commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -362,6 +362,17 @@ define Device/tplink_cpe610-v2
+define Device/tplink_cpe710-v1
+  $(Device/tplink-safeloader-uimage)
+  SOC := qca9563
+  IMAGE_SIZE := 15680k
```
Should this be: firmware size 0xf40000 / 1024 = 15616k ?

apcameron replied:
@jwmullally The Firmware size is actually 0xf50000 which is 15680k
Refer to the DTS target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts

jwmullally replied:
Is there a mismatch between the DTS and tplink-safeloader.c partition list?
```
			{"device-id", 0x63000, 0x00100},
			{"firmware", 0x70000, 0xf40000},
			{"soft-version", 0xfb0000, 0x00100},
			{"support-list", 0xfb1000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
```

```
			info: partition@60000 {
				label = "info";
				reg = <0x060000 0x010000>;
				read-only;
			};
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf50000>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfc0000 0x030000>;
				read-only;
			};
```

svanheule replied:
@jwmullally, see the discussion at https://github.com/openwrt/openwrt/pull/4143#issuecomment-841177042
I believe this mismatch is intentional, and used to reclaim some flash space for use by the rootfs partition.

jwmullally replied:
Rearranging the non-persistent partitions is fine.

Having `firmware` have different sizes in tplink-safeloader and the DTS/Makefile just to overlap `soft-version` and `support-list` like this is basically taking advantage of edge-case undefined behaviour in tplink-safeloader, TP-Link's OEM firmware upgrade and sysupgrade to get 64k (0.4%) more space. I don't think its worth the risk, and will just result in ongoing confusion and bugs. Has it been tested to the full firmware size?

I think its much better to stick with a standard predictable partition layout where nothing overlaps and everything is aligned. 

Either (preferable):
```
# tplink-safeloader.c
			{"firmware", 0x70000, 0xf40000},
			{"soft-version", 0xfb0000, 0x00100},
			{"support-list", 0xfb1000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
# DTS
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf50000>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfc0000 0x030000>;
				read-only;
			};
```
or much messier (might not be possible to due to alignment issues? Not worth it for 59k imho):
```
# tplink-safeloader.c
			{"firmware", 0x70000, 0xf4ef00},
			{"soft-version", 0xfbef00, 0x00100},
			{"support-list", 0xfbf000, 0x01000},
			{"user-config", 0xfc0000, 0x10000},
# DTS
			partition@70000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x070000 0xf4ef00>;
			};
			partition@fc0000 {
				label = "config";
				reg = <0xfbf000 0x031100>;
				read-only;
			};
```

If this overlap is being kept, the firmware section of the tplink-safeloader.c and DTS should include comments explaining whats going on.

jwmullally replied:
@svanheule Ok, I see now this is historically how it works for *all* the 15 existing CPE/WBS devices, and none of the other 46 TP-Link units, so this discussion probably belongs in another PR like [the one you linked above](https://github.com/openwrt/openwrt/pull/4143).

Okay, you can ignore this particular thread, sorry for the distraction :)

```
$ grep "\.id\|\.last_sysupgrade_partition" tools/firmware-utils/src/tplink-safeloader.c 
		.id     = "CPE210",
		.last_sysupgrade_partition = "support-list",
		.id     = "CPE210V2",
		.last_sysupgrade_partition = "support-list",
...
		.id     = "RE500-V1",
		.last_sysupgrade_partition = "file-system"
		.id     = "RE650-V1",
		.last_sysupgrade_partition = "file-system"
```

jwmullally replied:
> For the CPE and WBS devices sysupgrade back to the OEM Firmware generally does not work.

FYI, for this unit, I tested removing the `firmware` DTS -> `soft-version` `support-list` +64k hack (i.e. DTS firmware 0xf40000) and sysupgrade using OEM firmware works again.

Its up to ye, but I would recommend using a non-overlapping "standard" TP-Link partition layout for this unit (i.e. no hack) to put this in "right", and explore separately if this fixes the issue for the other CPE units.

svanheule replied:
As mentioned in https://github.com/openwrt/openwrt/pull/4117 the runtime partition layout may not be identical to the one stored in the partition table. On the other CPE units, the order of the partitions was also modified. Maybe that's the reason sysupgrade-to-stock is broken there, because an incompatible partition layout is used?

jwmullally replied:
@svanheule Yes, due to fixed `mtdparts`, OEM sysupgrade may not be guaranteed to work (e..g rootfs would be misaligned by 64k), although it does actually work in this case.
I agree that TFTP should still be the only recommended way of flashing, with OEM sysupgrade as an unsupported nice-to-have expert feature (if it fails, they can just fall back to TFTP). In this case, it could save people taking their unit down from the antennae.

(There a many improvements that could be made to `tplink-safeloader -z oem.bin`. `convert_firmware()` is currently very simple and produces only a raw contiguous output of `os-image`  + `file-system`, without considering if there is anything is in between. It doesn't add safety metadata. It could use an md5sum whitelist of tested OEM firmwares etc).

The overall point of this thread though is just simplifying this area (and not blocking oem sysupgrade) by removing the overlap for 16MB CPE/WBS units.

jwmullally commented:
I tested this on my own unit, and everything seems to be working OK :)

Re: reverting back to stock:

I tried to restore using a sysupgrade image generated from the OEM firmware and tplink-safeloader, but it bricked the unit.

```bash
openwrt-imagebuilder-21.02.0-rc1-ath79-generic.Linux-x86_64/staging_dir/host/bin/tplink-safeloader -z CPE710\(UN\)_1.0-up-ver2-2-0-P1\[20200304-rel70345\].bin -o sysupgrade-oem.bin
```

Fortunately, TFTP restore works OK using these official instructions: <https://www.tp-link.com/en/support/faq/2954/>
Everything in that link is accurate, including the IP addresses, however I had to connect the unit and my tftp server to a switch as a direct Ethernet connection recommended by the article didn't work (maybe due to slow interface initialization in my laptop while the unit boots).

apcameron commented:
@jwmullally For the CPE and WBS devices sysupgrade back to the OEM Firmware generally does not work.
The correct Process is to use the method recommended by TP-Link however I have found it to work best if I use a switch as I have found that it does not always work first time when not using a switch. I have also been successful using a TFTP server on Linux but again using a switch there too works best.

adschm commented:
@jwmullally If you want to touch this firmware partition size issue, please discuss directly with NeoRaider as I did back then in the linked mail thread. Note that he is a committer.

jwmullally commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,142 @@
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
```
Does this work? This is the only tp-link ath79 device with such a partition defined. Does it need the read-only removed to write to it?

jwmullally commented on target/linux/ath79/dts/qca9563_tplink_cpe710-v1.dts:
```diff
@@ -0,0 +1,142 @@
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
```
Does this work? This is the only tp-link ath79 device with such a partition defined. Does it need the read-only removed to write to it?

apcameron replied:
Yes it works.
I personally don't see why we would need to write to this partition. 
 I have it defined this way to match the OEM way of doing it 
The Entry from the OEM boot log is
[    0.980000] m25p80 spi0.0: found gd25q128, expected m25p80
[    0.990000] m25p80 spi0.0: gd25q128 (16384 Kbytes)
[    0.990000] 8 cmdlinepart partitions found on MTD device spi0.0
[    1.000000] Creating 8 MTD partitions on "spi0.0":
[    1.010000] 0x000000000000-0x000000040000 : "u-boot"
**_[    1.010000] 0x000000040000-0x000000050000 : "u-boot-env"_**
[    1.020000] 0x000000050000-0x000000070000 : "info"
.....

jwmullally replied:
Re removing the read-only flag, see the end of https://openwrt.org/docs/techref/bootloader/uboot.config
I guess getting it properly configured could be done in seperate PRs.

mpratt14 replied:
if there is a good reason to make uboot-env writable, it would be handled in this PR

I don't see the need, but I usually do this for boards I add whether it's needed or not

adschm replied:
u-boot-env should only be writable if there is a reason for it (typically uboot-envtools entry etc.).

apcameron replied:
@jwmullally I have removed the read-only from u-boot-env partition. Please test with your device.

apcameron commented:
@blogic Please review all the comments and let us know if you feel any changes are in order based on your testing and the comments in this PR or if you feel that it is ready to be committed.

apcameron commented:
Now that I have added _serial-number_ to definition in _tools/firmware-utils/src/tplink-safeloader.c_ and
removed  _read-only_ from the _u-boot-env_ partition in my last commit  is the PR ready to be Merged to master and possibly backported to the 21.02 release or does it need further changes?

It has been tested at various stages by myself, @blogic and @jwmullally

apcameron commented:
@adschm @mpratt14 Do I need To make any more changes or is the PR Ready?

apcameron commented:
@adschm  @mpratt14 I have just noticed something strange.
If I have the code for my PR applied and then build an image for my TP-LINK CPE610 V1 I see the following which is not correct.
Why is the ath10k folder being created with firmware that is unrelated to the TP-LINK CPE610 v1?

<pre>root@OpenWrt:/tmp/sysinfo# cd /lib/firmware/
root@OpenWrt:/lib/firmware# ls
<font color="#3465A4"><b>ath10k</b></font>         regulatory.db
root@OpenWrt:/lib/firmware# ls -al
drwxr-xr-x    3 root     root            50 Jun 23 22:21 <font color="#3465A4"><b>.</b></font>
drwxr-xr-x   11 root     root           485 Jun 23 22:21 <font color="#3465A4"><b>..</b></font>
drwxr-xr-x    3 root     root            30 Jun 23 22:21 <font color="#3465A4"><b>ath10k</b></font>
-rw-r--r--    1 root     root          4000 Jun 23 22:21 regulatory.db
root@OpenWrt:/lib/firmware# cat /tmp/sysinfo/*
tplink,cpe610-v1
TP-Link CPE610 v1
root@OpenWrt:/lib/firmware# 
</pre>

mpratt14 commented:
just a guess because not much info provided...

if you built for multiple devices including one that selects ath10k drivers, and without selecting
`CONFIG_TARGET_PER_DEVICE_ROOTFS`, then it will use the same rootfs for each image
and this means all images get all selected packages from all selected devices

apcameron commented:
@mpratt14 Thank you that explains it as I did not have CONFIG_TARGET_PER_DEVICE_ROOTFS set

apcameron commented:
@adschm Is there any chance of getting this added/backported to the 21.02 release/branch?

adschm commented:
> @adschm Is there any chance of getting this added/backported to the 21.02 release/branch?

It's not even merged yet.

apcameron commented:
@adschm Does it need any further changes or is it ready to be merged?

tiagofreire-pt commented:
Just to assure this PR is quite relevant for the community, opening Openwrt to a new device aiming long distance complex infrastructures and getting rid of buggy and unsafe OEM firmware.

Could the reviewers give some attention on this PR, could they? :)

apcameron commented:
@adschm Rebased the code and updated .soft_ver = SOFT_VER_DEFAULT,

adschm commented:
Thanks, I've converted this to nvmem MAC address assignment and will merge it during the weekend.

I assume MAC addresses for ethernet and WiFi are the same? (Since you have an incomplete macaddr_add in 10-ath10k-caldata)

tiagofreire-pt commented:
> 
> I assume MAC addresses for ethernet and WiFi are the same? (Since you have an incomplete macaddr_add in 10-ath10k-caldata)

Yes, they are. At least with the original firmware on client mode.

robimarko commented:
I dont think that you will see devices backported to a stable release, that would kind of break the purpose of a stable release

NeoRaider commented:
As long as a new device can be backported without changing existing devices significantly, I generally support backports.

adschm commented:
> @adschm can the commit from this PR also be merged unchanged as a backport to `/openwrt-21.02`, or should that be done with a new PR?

https://openwrt.org/docs/guide-developer/device-support-policies#backports

Judging whether that's possible requires keeping track of all the relevant changes. I personally prefer to invest my time in features for master for the moment, and thus cannot do this judgement properly. The problem is not doing the changes, but being sure about whether (and which) changes are necessary.

Of course, if somebody else wants to invest time into reviewing a feature backport, I'm perfectly fine with that.
